src/components/2FA/helpers.js

```diff
@@ -0,0 +1,132 @@
+// encrypted with the encryption key, and the public key encoded in base64.
+function makeKeyPair(symKey) {
+  const subtle = w.crypto.subtle
+
+  const encKey = symKey.slice(0, 32)
+  const macKey = symKey.slice(32, 64)
+  const iv = randomBytes(16)
+  const rsaParams = {
+    name: 'RSA-OAEP',
+    modulusLength: 2048,
+    publicExponent: new Uint8Array([0x01, 0x00, 0x01]), // 65537
+    hash: { name: 'SHA-1' }
+  }
+  const hmacParams = { name: 'HMAC', hash: 'SHA-256' }
+  let publicKey, privateKey, encryptedKey
+  return subtle
+    .generateKey(rsaParams, true, ['encrypt', 'decrypt'])
+    .then(pair => {
+      const publicPromise = subtle.exportKey('spki', pair.publicKey)
+      const privatePromise = subtle.exportKey('pkcs8', pair.privateKey)
+      return Promise.all([publicPromise, privatePromise])
+    })
+    .then(keys => {
+      publicKey = keys[0]
+      privateKey = keys[1]
+      return subtle.importKey('raw', encKey, { name: 'AES-CBC' }, false, [
+        'encrypt'
+      ])
+    })
+    .then(impKey =>
+      subtle.encrypt({ name: 'AES-CBC', iv: iv }, impKey, privateKey)
+    )
+    .then(encrypted => {
+      encryptedKey = encrypted
+      return subtle.importKey('raw', macKey, hmacParams, false, ['sign'])
+    })
+    .then(impKey => {
+      const macData = new Uint8Array(iv.byteLength + encryptedKey.byteLength)
+      macData.set(new Uint8Array(iv), 0)
+      macData.set(new Uint8Array(encryptedKey), iv.byteLength)
+      return subtle.sign(hmacParams, impKey, macData)
+    })
+    .then(mac => {
+      const public64 = fromBufferToB64(publicKey)
+      const iv64 = fromBufferToB64(iv)
+      const priv = fromBufferToB64(encryptedKey)
+      const mac64 = fromBufferToB64(mac)
+      return {
+        publicKey: public64,
+        // 2 means AesCbc256_HmacSha256_B64
+        privateKey: `2.${iv64}|${priv}|${mac64}`
+      }
+    })
+}
+
+export async function forceSetPassphrase(client, passphrase, hint) {
+  const instanceURL = client.getStackClient().uri
+  const vaultClient = new WebVaultClient(instanceURL)
+
+  let { Kdf: kdf, KdfIterations: kdfIterations } =
+    await client.stackClient.fetchJSON(
+      'POST',
+      '/bitwarden/api/accounts/prelogin'
+    )
+
+  if (kdfIterations === 0) {
+    kdfIterations = 650000
+  }
+
+  const masterKey = await vaultClient.computeMasterKey(
+    passphrase,
+    kdfIterations,
+    kdf
+  )
+  const passwordHash = await vaultClient.computeHashedPassword(
+    passphrase,
+    masterKey
+  )
+
+  const encryptionKey = await makeEncKey(masterKey.encKey)
+  const publicPrivateKeys = await makeKeyPair(encryptionKey.key)
+
+  await client.stackClient.fetchJSON('PUT', '/settings/passphrase', {
+    new_passphrase: passwordHash,
+    hint,
+    key: encryptionKey.cipherString,
+    publicKey: publicPrivateKeys.publicKey,
+    privateKey: publicPrivateKeys.privateKey,
+    iterations: 650000,
+    force: true
+  })
+}
```

Comment:
@Ldoppea Don't we have shared code for that?

Reply:
From the discussion I had with him we don't have a shared code for that at the moment. We need to start a conversation with @nono as well because the code is coming from the stack initially to make sure we do the same thing everywhere